lib/internal/modules/esm/loader.js

```diff
@@ -202,33 +211,74 @@ class ESMLoader {
       const { ModuleWrap, callbackMap } = internalBinding('module_wrap');
       const module = new ModuleWrap(url, undefined, source, 0, 0);
       callbackMap.set(module, {
-        importModuleDynamically: (specifier, { url }) => {
-          return this.import(specifier, url);
+        importModuleDynamically: (specifier, { url }, importAssertions) => {
+          return this.import(specifier, url, importAssertions);
         }
       });
 
       return module;
     };
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
+    finalFormatCache.set(job, 'module');
     const { module } = await job.run();
 
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async getModuleJob(specifier, parentURL) {
+  async getModuleJob(specifier, parentURL, importAssertions) {
+    if (!ArrayPrototypeIncludes(supportedTypes, importAssertions.type)) {
+      throw new ERR_INVALID_IMPORT_ASSERTION('type', importAssertions.type);
+    }
+
     const { format, url } = await this.resolve(specifier, parentURL);
     let job = this.moduleMap.get(url);
     // CommonJS will set functions for lazy job evaluation.
     if (typeof job === 'function') this.moduleMap.set(url, job = job());
 
-    if (job !== undefined) return job;
+    if (job != null) {
+      const currentImportAssertionType = importAssertionTypeCache.get(job);
+      if (currentImportAssertionType === importAssertions.type) return job;
+
+      try {
+        // To avoid race conditions, wait for previous module to fulfill first.
+        await job.modulePromise;
+      } catch {
+        // If the other job failed with a different `type` assertion, we got
+        // another chance.
+        job = undefined;
+      }
+
+      if (job !== undefined) {
+        const finalFormat = finalFormatCache.get(job);
+        if (importAssertions.type == null && finalFormat === 'json') {
+          throw new ERR_MISSING_IMPORT_ASSERTION(url, finalFormat,
+                                                 'type', 'json');
+        }
+        if (
+          importAssertions.type == null ||
+              (importAssertions.type === 'json' && finalFormat === 'json')
+        ) return job;
+        throw new ERR_FAILED_IMPORT_ASSERTION(
+          url, 'type', importAssertions.type, finalFormat);
+      }
+    }
 
     const moduleProvider = async (url, isMain) => {
       const { format: finalFormat, source } = await this.load(url, { format });
 
+      if (importAssertions.type === 'json' && finalFormat !== 'json') {
+        throw new ERR_FAILED_IMPORT_ASSERTION(
+          url, 'type', importAssertions.type, finalFormat);
+      }
+      if (importAssertions.type !== 'json' && finalFormat === 'json') {
+        throw new ERR_MISSING_IMPORT_ASSERTION(url, finalFormat,
+                                               'type', 'json');
+      }
+      finalFormatCache.set(job, finalFormat);
```

Comment:
With a one-to-one relationship, I don’t think we need `finalFormatCache` or `importAssertionTypeCache`, do we?

Reply:
We do, or at least we need something to handle the race condition case. But open to other suggestions as long as the tests pass.

Reply:
where is it clearing / which cache?

Reply:
Line 303
```js
this.moduleMap.set(url, job);
```

Reply:
When would there ever be a mismatch (the `null`/`"json"` or `"json"`/`null` cases in the table above)? I would think that if assertions can never be optional (as in, always required for JSON and never required for other formats) then it wouldn’t be possible for the cache to contain an entry that has a different assertion than what is being requested. Or put another way, if `file.json` is in the module cache, it had to have had `assert: { type: 'json' }` or else it would’ve errored before getting into the cache. Likewise if `file.js` is in the module cache, it couldn’t have had an assertion or else _that_ `import` would have errored.

If I’m correct about this, then these last two cases don’t exist; and then I would think that we don’t need the `WeakMap`s, and `importAssertions` doesn’t need to be an argument of `getModuleJob`.

Reply:
It’s not yet decided tho that they can never be optional, for json or for unknown future module formats (this remains a possibility on the web, too) - shouldn’t things be built without hardcoding that assumption?

Reply:
> When would there ever be a mismatch (the null/"json" or "json"/null cases in the table above)?

Check the test cases:

```js
Promise.all([
  import('specifier'),
  import('specifier', { assert: {type: 'json' }}),
])
Promise.all([
  import('specifier', { assert: {type: 'json' }}),
  import('specifier'),
])
```

> I would think that if assertions can never be optional (as in, always required for JSON and never required for other formats) then it wouldn’t be possible for the cache to contain an entry that has a different assertion than what is being requested.

Well I don't think that's true, and tbh, I don't understand why you would think that: the loader cannot know until the `load` hook has resolved what format a specifier is to check if the import assertion is met.
So if `file.json` or `file.js` is in the module cache, until we know what format it is, we cannot conclude what to do with the new request.So yeah, the cache cannot be a 1-to-1 relationship url <-> module job anymore, and I don't think there's a way around it. Check out https://tc39.es/proposal-import-assertions/#sec-hostresolveimportedmodule for more info, and the whole discussion in https://github.com/nodejs/node/pull/39921 as it was discussed extensively there.

Let's not digress discussing assertionless imports for now.

Reply:
> Check the test cases:

Which test cases are you referring to? I don’t see any that quite match what you’re describing here.

In the current code of this PR, `this.moduleMap.set(url, job);` isn’t called until the very end of `getModuleJob`, after all the validation checks. So an `import` that fails validation never goes into the module cache, because an exception would be thrown before we’d get to `this.moduleMap.set`. Which means, I think, that there’s no way to create a mismatch like the `null`/`"json"` or `"json"`/`null` cases, is there?

Reply:
> Which test cases are you referring to?

https://github.com/nodejs/node/pull/40250/files#diff-2e85145c8ebda76ab69a44c813827185e6496e6b2631f73c0e84473182cfd949R27-R45

> In the current code of this PR, this.moduleMap.set(url, job); isn’t called until the very end of getModuleJob, after all the validation checks.

Actually the validation checks do not happen in `getModuleJob` but in the `moduleProvider` function, after the `await this.load(url)`, while the `this.moduleMap.set(url, job)` is made synchronously.

As I told you by email, I don't feel the conversation through comments is very productive,  and like Bradley said in https://github.com/nodejs/node/pull/39921#pullrequestreview-762214543, I think it’s time to take that discussion to a meeting.